### PR TITLE
Add client config_id abstraction

### LINKED ISSUE #76: Client config_id abstraction
Currently the client gets th id from for the db. There are no permission checks and clients could potentially guess ids.
Ideally we need a per-client hashmap that maps a client config_id to the corresponding db config_id and information about it's permissions to the config (readonly / readwrite).

### PR DESCRIPTION
**Pull request by Milerius**
Closes #76

### CODE REVIEW

File: service/service.hpp [hunk 1]
```diff
@@ -199,10 +208,11 @@ namespace raven
 
     void include_config(json::json &json_data, uvw::PipeHandle &sock)
     {
+        //TODO: Communication.md has changed, need to be modified
```
Comment (Gaspard--) [inline service/service.hpp]: Please open as issue in that case.

File: albinos-cli/source/protocol/protocol_type.d [hunk 2]
```diff
@@ -26,5 +26,5 @@ struct config_load_answer
 {
     @serializationKeys("REQUEST_STATE") string state;
     @serializationKeys("CONFIG_NAME") string config_name;
-    @serializationKeys("CONFIG_ID") int config_id;
```
Comment (Gaspard--) [inline albinos-cli/source/protocol/protocol_type.d]: Why the change ?

Comment (Milerius) [reply albinos-cli/source/protocol/protocol_type.d]: Config id is a size_t now

Comment (Gaspard--) [reply albinos-cli/source/protocol/protocol_type.d]: but why would you want to switch ?

Comment (Arignir) [reply albinos-cli/source/protocol/protocol_type.d]: btw, `sitez_t` != `ulong`. Such assumption is not portable :(

Comment (Milerius) [reply albinos-cli/source/protocol/protocol_type.d]: @Arignir Yes. Fixed, thank's

@Gaspard-- std::hash<string>() return a size_t. That's why i switch int to size_t, config_id is now a size_t, and error are handler through a std::optional<request_state>, double check the review

File: service/service.hpp [hunk 3]
```diff
@@ -164,15 +164,19 @@ namespace raven
     {
         auto cfg = fill_request<config_create>(json_data);
         auto config_create_db_result = db_.config_create(json_data);
-        if (config_create_db_result.config_id.value() != static_cast<int>(request_state::db_error)) {
-            config_clients_registry_.at(sock.fileno()) += config_create_db_result.config_id;
+        if (!config_create_db_result.state_error.has_value()) {
+            using namespace std::string_literals;
+            auto unique_id = std::hash<std::string>()(cfg.config_name + "_"s + std::to_string(sock.fileno()) + "_"s +
+                std::to_string(config_create_db_result.config_id.value()));
+            config_clients_registry_.at(sock.fileno()) += std::make_pair(config_id_st{unique_id},
```
Comment (Gaspard--) [inline service/service.hpp]: Why do you do the association in create_config ? Shouldn't this happen in load_config ?

Comment (Milerius) [reply service/service.hpp]: No, we need to associate it, internally, config_create return the db id, this is perfect to make the association with the unique id.

Comment (Riz-cros) [reply service/service.hpp]: this isn't completly related to this matter, but I wonder if create config should also load the config once it's created ?

Comment (Gaspard--) [reply service/service.hpp]: Well, this doesn't work if the config is loaded when it wasn't created in this run of the service.
@Riz-cros Create_config should not load the config, it does not return the config_id.

**Comment by Gaspard--** [comment]
Please provide more information as to what exactly this PR is trying to achieve, I think I may have misunderstood what you are trying to do.

File: service/service.hpp (hunk 3 above)
Comment (Riz-cros) [reply service/service.hpp]: So the unique id is the key in that case ?

Comment (Riz-cros) [reply service/service.hpp]: So the unitque id is the key in that case ?

File: service/service_strong_types.hpp [hunk 4]
```diff
@@ -9,7 +9,7 @@
 
 namespace raven
 {
-  using config_id_st = st::type<int, struct config_id_tag, st::arithmetic>;
+  using config_id_st = st::type<std::size_t, struct config_id_tag, st::arithmetic>;
```
Comment (Gaspard--) [inline service/service_strong_types.hpp]: I think we can safely assume the client will load less than 2^32 configs. You could even strengthen that assumption to 2^16 or even 2^8 (that last one may be too far though).

File: service/client.hpp [hunk 5]
```diff
@@ -22,19 +22,26 @@ namespace raven
...
 
+    config_id_st get_current_id() const noexcept
+    {
+        return current_id;
+    }
+
   private:
     client_ptr sock_;
-    std::unordered_set<raven::config_id_st::value_type> config_ids_;
+    raven::config_id_st current_id{0};
```
[the remark this answers is not in the document]
Comment (Riz-cros) [reply service/client.hpp]: Indeed, at first glance I didn't understand what the current_id meant, next_id seems the best to me

[the remark this answers is not in the document]
Comment (Riz-cros) [reply service/client.hpp]: actually something more like `last_id` would be better since, adding a config id increment the "curent_id" then insert the config, it make sense to call for a function afterward to get the last id stored
